Observing all readers failed in longevity run with "DataCorruptionException"

**Problem description**

All reader (0/8) failure observed while running IO workload of ~6 Mbps having readers: 8, writers: 4, scaleFactor: 2, maxEventsPerSecond: 1000, targetRate: 200, scalingType: EVENTS_PER_SECOND for 1+ days. Since all the readers have failed so there is no event population for readers. Writers did not fail (4/4) but the event rate is very low (~1000 events/sec), which was (~3000 events/sec) in the start.

In client manager-log during first reader failure (7/8) observed `RetriesExhaustedException` on `segment=longevity/mediumScale/1001.#epoch.646`, following is the log from that instance : -
```
INFO  [2019-05-28 21:16:59,690] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:195,558,248, events/sec:3,969, KB/sec:3043.7207
INFO  [2019-05-28 21:16:59,690] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-05-28 21:16:59,690] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 0, average block time (ms): 0
INFO  [2019-05-28 21:16:59,691] io.pravega.longevity.utils.PerformanceUtils: Readers (8/8): events:195,551,721, events/sec:3,969, KB/sec:3043.6191
172.24.37.2 - - [28/May/2019:21:16:59 +0000] "POST /v1/longevity/performance HTTP/1.1" 200 2720 "-" "Jersey/2.25.1 (HttpUrlConnection 1.8.0_131)" 12
INFO  [2019-05-28 21:17:04,002] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=27577985007616, segment=longevity/mediumScale/1001.#epoch.646, correctHost=, serverStackTrace=)
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=27577985007616, segment=longevity/mediumScale/1001.#epoch.646, correctHost=, serverStackTrace=)
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:121)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:94)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=27577985007616, segment=longevity/mediumScale/1001.#epoch.646, correctHost=, serverStackTrace=)
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.failAllInflight(AsyncSegmentInputStreamImpl.java:265)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.closeConnection(AsyncSegmentInputStreamImpl.java:237)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.access$100(AsyncSegmentInputStreamImpl.java:42)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.wrongHost(AsyncSegmentInputStreamImpl.java:70)
        at io.pravega.shared.protocol.netty.WireCommands$WrongHost.process(WireCommands.java:125)
        at io.pravega.shared.protocol.netty.ReplyProcessor.process(ReplyProcessor.java:20)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelRead$2(FlowHandler.java:209)
        at java.util.Optional.ifPresent(Unknown Source)
        at io.pravega.client.netty.impl.FlowHandler.channelRead(FlowHandler.java:207)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        ... 1 more
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=27577985007616, segment=longevity/mediumScale/1001.#epoch.646, correctHost=, serverStackTrace=)
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
        ... 39 more
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=27577985007616, segment=longevity/mediumScale/1001.#epoch.646, correctHost=, serverStackTrace=)
        ... 34 more
172.24.37.2 - - [28/May/2019:21:17:04 +0000] "POST /v1/longevity/error HTTP/1.1" 200 6971 "-" "Jersey/2.25.1 (HttpUrlConnection 1.8.0_131)" 324
INFO  [2019-05-28 21:17:29,733] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 13h 42m
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:195,558,248, events/sec:3,965, KB/sec:3040.0186
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 0, average block time (ms): 0
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: Readers (7/8): events:195,555,720, events/sec:3,965, KB/sec:3039.9785
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: New runtime for 8a9c3312-b2be-4d31-853a-76e6c79ab402  Running for 13h 42m
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: Dumping stats for stream longevity/mediumScale
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:195,558,248, events/sec:3,965, KB/sec:3040.0186
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 0, average block time (ms): 0
INFO  [2019-05-28 21:17:29,734] io.pravega.longevity.utils.PerformanceUtils: Readers (7/8): events:195,555,720, events/sec:3,965, KB/sec:3039.9785

```

In segment store found  `segment=longevity/mediumScale/1001.#epoch.646`, observed `ERROR i.p.s.server.writer.StorageWriter - StorageWriter[5]: Iteration[40].CriticalError` and `ERROR i.p.s.s.s.StreamSegmentContainerRegistry - Critical failure for SegmentContainer` along with `io.pravega.segmentstore.server.DataCorruptionException` in Segment Store.
Following are the exceptions from segment store: -
```
2019-05-28 21:15:35,769 114639189 [core-17] INFO  i.p.s.server.CacheManager - CacheManager Registered AttributeIndex[102458].
2019-05-28 21:15:35,781 114639201 [core-23] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[5]: Iteration[40].CriticalError.
io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 102403, Offsets = [461052-461160), SeqNo = 770486'.
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:826)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:696)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:712)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:681)
        at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:572)
        at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:725)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2019-05-28 21:15:35,781 114639201 [core-23] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[5]: Stopping.
2019-05-28 21:15:35,781 114639201 [core-23] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[5-102401]: State changed from Writing to Closed.
2019-05-28 21:15:35,781 114639201 [core-23] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[5-102400]: State changed from Writing to Closed.
```

```
2019-05-28 21:15:35,839 114639259 [core-29] INFO  i.p.s.s.s.StreamSegmentContainerRegistry - Unregistered SegmentContainer 5.
2019-05-28 21:15:35,839 114639259 [core-29] ERROR i.p.s.s.s.StreamSegmentContainerRegistry - Critical failure for SegmentContainer Container Id = 5, State = FAILED. {}
io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 102403, Offsets = [461052-461160), SeqNo = 770486'.
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:826)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:696)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:712)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:681)
        at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:572)
        at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:725)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2019-05-28 21:15:35,975 114639395 [epollEventLoopGroup-11-6] WARN  i.p.s.s.h.handler.AppendProcessor - Wrong host. Segment 'Scope/test1/97.#epoch.0' (Container 5) is not owned and 84ffb4dd-e22f-436b-b076-cab51c3e253f cannot perform operation 'appending data'.
```

Pravega-Operator deployment with 2 controllers, 3 Zookeepers, 4 Bookies and 3 Segment stores.
Environment details: PKS / K8 with medium cluster:
```
1 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
3 worker nodes @ xlarge.cpu(4 cpu, 16 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Pravega details:
```
Pravega version : 0.5.0-2260.b477eec
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.2
```

Observing one of the container component failure warning messages in segment store : -
```
019-05-28 21:15:35,782 114639202 [core-23] INFO  i.p.s.s.writer.SegmentAggregator - StorageWriter[5-102398]: State changed from Writing to Closed.
2019-05-28 21:15:35,782 114639202 [core-17] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[5]: Writer failed. Shutting down StreamSegmentContainer.
java.util.concurrent.CompletionException: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 102403, Offsets = [461052-461160), SeqNo = 770486'.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:975)
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:697)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:712)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:681)
        at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:572)
        at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:725)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.pravega.segmentstore.server.DataCorruptionException: Unable to retrieve CacheContents for 'AggregatedAppend: SegmentId = 102403, Offsets = [461052-461160), SeqNo = 770486'.
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:826)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:696)
        ... 11 common frames omitted
2019-05-28 21:15:35,782 114639202 [core-17] INFO  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[5]: Stopping.
2019-05-28 21:15:35,783 114639203 [core-17] INFO  i.p.c.c.AbstractThreadPoolService - SegmentContainer[5]: Stopping.
2019-05-28 21:15:35,783 114639203 [core-17] INFO  i.p.c.c.AbstractThreadPoolService - SegmentContainer[5]: Stopped.
2019-05-28 21:15:35,783 114639203 [core-17] INFO  i.p.s.server.logs.DurableLog - DurableLog[5]: Stopping.
2019-05-28 21:15:35,783 114639203 [core-17] INFO  i.p.c.c.AbstractThreadPoolService - OperationProcessor[5]: Stopping.
2019-05-28 21:15:35,783 114639203 [core-17] INFO  i.p.c.c.AbstractThreadPoolService - OperationProcessor[5]: Stopped.
2019-05-28 21:15:35,782 114639202 [core-23] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[5]: Stopped.
2019-05-28 21:15:35,786 114639206 [core-17] INFO  i.p.s.s.i.bookkeeper.BookKeeperLog - Log[5]: Closed.
2019-05-28 21:15:35,786 114639206 [core-17] INFO  i.p.s.server.logs.DurableLog - DurableLog[5]: Stopped.
2019-05-28 21:15:35,786 114639206 [core-29] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[5]: Failed due to component failure.
2019-05-28 21:15:35,787 114639207 [core-29] INFO  i.p.s.server.CacheManager - CacheManager Unregistered io.pravega.segmentstore.server.tables.ContainerKeyCache@2f62aace.
2019-05-28 21:15:35,788 114639208 [core-29] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_5_TableKeys]: Closed.
2019-05-28 21:15:35,788 114639208 [core-29] INFO  i.p.c.c.AbstractThreadPoolService - SegmentContainer[5]: Closed.
2019-05-28 21:15:35,788 114639208 [core-29] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[5]: Closed.
2019-05-28 21:15:35,788 114639208 [core-29] INFO  i.p.c.c.AbstractThreadPoolService - OperationProcessor[5]: Closed.
2019-05-28 21:15:35,788 114639208 [core-29] INFO  i.p.s.server.logs.DurableLog - DurableLog[5]: Closed.
```

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMetadata.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMetadata.java
@@ -9,6 +9,7 @@
  */
 package io.pravega.segmentstore.server.containers;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 import io.pravega.common.Exceptions;
 import io.pravega.common.function.Callbacks;
@@ -355,7 +356,8 @@ public class StreamSegmentMetadata implements UpdateableSegmentMetadata {
     /**
      * Marks this SegmentMetadata as inactive.
      */
-    synchronized void markInactive() {
+    @VisibleForTesting
+    public synchronized void markInactive() {
         this.active = false;
     }
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/SegmentMetadataUpdateTransaction.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/SegmentMetadataUpdateTransaction.java
@@ -125,7 +125,7 @@ class SegmentMetadataUpdateTransaction implements UpdateableSegmentMetadata {
 
     @Override
     public boolean isActive() {
-        throw new UnsupportedOperationException("isActive() is not supported on " + getClass().getName());
+        return true;
     }
 
     @Override

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java
@@ -9,6 +9,7 @@
  */
 package io.pravega.segmentstore.server.reading;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 import io.pravega.common.Exceptions;
 import io.pravega.common.ObjectClosedException;
@@ -245,15 +246,16 @@ public class ContainerReadIndex implements ReadIndex {
             }
 
             for (long streamSegmentId : segmentIds) {
-                SegmentMetadata segmentMetadata = this.metadata.getStreamSegmentMetadata(streamSegmentId);
-                boolean wasRemoved = false;
-                if (segmentMetadata == null || segmentMetadata.isDeleted() || !segmentMetadata.isActive()) {
-                    wasRemoved = closeIndex(streamSegmentId, true);
-                }
-
-                if (wasRemoved) {
+                StreamSegmentReadIndex index = this.readIndices.get(streamSegmentId);
+                if (index != null && !index.isActive()) {
+                    // This index is registered, but its metadata indicates it is no longer active - close it.
+                    closeIndex(streamSegmentId, true);
+                    removed.add(streamSegmentId);
+                } else if (index == null) {
+                    // This index is not registered.
                     removed.add(streamSegmentId);
                 } else {
+                    // This index is registered, but its metadata indicates it is still active; don't do anything.
                     notRemoved.add(streamSegmentId);
                 }
             }
@@ -263,7 +265,7 @@ public class ContainerReadIndex implements ReadIndex {
             log.debug("{}: Unable to clean up ReadIndex for Segments {} because no such index exists or the Segments are not deleted.", this.traceObjectId, notRemoved);
         }
 
-        log.info("{}: Cleaned up ReadIndices for deleted Segments {}.", this.traceObjectId, removed);
+        log.info("{}: Cleaned up ReadIndices for {} inactive or deleted Segments.", this.traceObjectId, removed);
     }
 
     @Override
@@ -341,7 +343,8 @@ public class ContainerReadIndex implements ReadIndex {
      *
      * @param streamSegmentId The Id of the StreamSegment whose ReadIndex to get.
      */
-    private StreamSegmentReadIndex getIndex(long streamSegmentId) {
+    @VisibleForTesting
+    StreamSegmentReadIndex getIndex(long streamSegmentId) {
         synchronized (this.lock) {
             return this.readIndices.getOrDefault(streamSegmentId, null);
         }
@@ -358,12 +361,21 @@ public class ContainerReadIndex implements ReadIndex {
         synchronized (this.lock) {
             // Try to see if we have the index already in memory.
             index = getIndex(streamSegmentId);
+            if (index != null && !index.isActive()) {
+                // Index is registered, but it points to a segment metadata that is inactive. We should not be using
+                // it anymore.
+                closeIndex(streamSegmentId, true);
+                index = null;
+            }
+
             if (index == null) {
-                // We don't have it, create one.
+                // Create a new Segment Read Index.
                 SegmentMetadata segmentMetadata = this.metadata.getStreamSegmentMetadata(streamSegmentId);
-                Exceptions.checkArgument(segmentMetadata != null, "streamSegmentId",
-                        "StreamSegmentId %s does not exist in the metadata.", streamSegmentId);
-                if (segmentMetadata.isDeleted()) {
+                if (segmentMetadata == null) {
+                    throw new IllegalArgumentException(String.format("Segment Id %d does not exist in the metadata.", streamSegmentId));
+                } else if (!segmentMetadata.isActive()) {
+                    throw new IllegalArgumentException(String.format("Segment Id %d does exist in the metadata but is inactive.", streamSegmentId));
+                } else if (segmentMetadata.isDeleted()) {
                     throw new StreamSegmentNotExistsException(segmentMetadata.getName());
                 }
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java
@@ -275,6 +275,15 @@ class StreamSegmentReadIndex implements CacheManager.Client, AutoCloseable {
     }
 
     /**
+     * Gets a value indicating whether the Segment that this ReadIndex points to is still active (in memory) and not deleted.
+     *
+     * @return True if active, false otherwise.
+     */
+    boolean isActive() {
+        return this.metadata.isActive() && !this.metadata.isDeleted();
+    }
+
+    /**
      * Gets the length of the Segment this ReadIndex refers to.
      */
     long getSegmentLength() {

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/reading/ContainerReadIndexTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/reading/ContainerReadIndexTests.java
@@ -19,11 +19,13 @@ import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
 import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
 import io.pravega.segmentstore.contracts.StreamSegmentTruncatedException;
 import io.pravega.segmentstore.server.CachePolicy;
+import io.pravega.segmentstore.server.EvictableMetadata;
 import io.pravega.segmentstore.server.MetadataBuilder;
 import io.pravega.segmentstore.server.SegmentMetadata;
 import io.pravega.segmentstore.server.TestCacheManager;
 import io.pravega.segmentstore.server.UpdateableContainerMetadata;
 import io.pravega.segmentstore.server.UpdateableSegmentMetadata;
+import io.pravega.segmentstore.server.containers.StreamSegmentMetadata;
 import io.pravega.segmentstore.storage.Cache;
 import io.pravega.segmentstore.storage.CacheFactory;
 import io.pravega.segmentstore.storage.Storage;
@@ -40,6 +42,7 @@ import java.io.InputStream;
 import java.io.UncheckedIOException;
 import java.time.Duration;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
@@ -773,10 +776,10 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
      * Tests the ability to evict entries from the ReadIndex under various conditions:
      * * If an entry is aged out
      * * If an entry is pushed out because of cache space pressure.
-     * <p>
+     *
      * This also verifies that certain entries, such as RedirectReadIndexEntries and entries after the Storage Offset are
      * not removed.
-     * <p>
+     *
      * The way this test goes is as follows (it's pretty subtle, because there aren't many ways to hook into the ReadIndex and see what it's doing)
      * 1. It creates a bunch of segments, and populates them in storage (each) up to offset N/2-1 (this is called pre-storage)
      * 2. It populates the ReadIndex for each of those segments from offset N/2 to offset N-1 (this is called post-storage)
@@ -949,6 +952,58 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
             //1.25N..1.5N
             checkOffsets(segmentRemovedKeys, segmentId, entriesPerSegment + entriesPerSegment / 4, entriesPerSegment / 4, (int) (entriesPerSegment * appendSize * 1.25), appendSize);
         }
+    }
+
+    /**
+     * Tests the {@link ContainerReadIndex#cleanup} method as well as its handling of inactive segments.
+     */
+    @Test
+    public void testCleanup() throws Exception {
+        // Create all the segments in the metadata.
+        @Cleanup
+        TestContext context = new TestContext();
+        ArrayList<Long> segmentIds = createSegments(context);
+
+        final long activeSegmentId = segmentIds.get(0); // Always stays active.
+        final long inactiveSegmentId1 = segmentIds.get(1); // Becomes inactive - used for cleanup()
+        final long inactiveSegmentId2 = segmentIds.get(2); // Becomes inactive - used for getOrCreateIndex() (any op).
+        final long reactivatedSegmentId1 = segmentIds.get(3); // Becomes inactive and then active again (for cleanup)
+        final long reactivatedSegmentId2 = segmentIds.get(4); // Becomes inactive and then active again (for getOrCreateIndex()).
+
+        // Add a zero-byte append, which ensures the Segments' Read Indices are initialized.
+        for (val id : segmentIds) {
+            context.readIndex.append(id, 0, new byte[0]);
+        }
+
+        // Mark 2 segments as inactive, but do not evict them yet. We simulate a concurrent eviction, when the segment is
+        // first marked as inactive and the Read Index receives a request for it before it gets evicted.
+        markInactive(inactiveSegmentId1, context);
+        markInactive(inactiveSegmentId2, context);
+
+        // Evict 2 segments (which also marks them as inactive), then re-map them as active segments (which gives them
+        // new instances of their Segment Metadatas).
+        val reactivatedSegment2OldIndex = context.readIndex.getIndex(reactivatedSegmentId2);
+        evict(reactivatedSegmentId1, context);
+        evict(reactivatedSegmentId2, context);
+        createSegment(reactivatedSegmentId1, context);
+        createSegment(reactivatedSegmentId2, context);
+
+        // Test cleanup().
+        context.readIndex.cleanup(Arrays.asList(activeSegmentId, inactiveSegmentId1, reactivatedSegmentId1));
+        Assert.assertNotNull("Active segment's index removed during cleanup.", context.readIndex.getIndex(activeSegmentId));
+        Assert.assertNull("Inactive segment's index not removed during cleanup.", context.readIndex.getIndex(inactiveSegmentId1));
+        Assert.assertNull("Reactivated segment's index not removed during cleanup.", context.readIndex.getIndex(reactivatedSegmentId1));
+
+        // Test getOrCreateIndex() via append() (any other operation could be used for this, but this is the simplest to setup).
+        AssertExtensions.assertThrows(
+                "Appending to inactive segment succeeded.",
+                () -> context.readIndex.append(inactiveSegmentId2, 0, new byte[0]),
+                ex -> ex instanceof IllegalArgumentException);
+
+        // This should re-create the index.
+        context.readIndex.append(reactivatedSegmentId2, 0, new byte[0]);
+        Assert.assertNotEquals("Reactivated Segment's ReadIndex was not re-created.",
+                reactivatedSegment2OldIndex, context.readIndex.getIndex(reactivatedSegmentId2));
     }
 
     // region Scenario-based tests
@@ -1367,7 +1422,7 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
         return segmentIds;
     }
 
-    private long createSegment(int id, TestContext context) {
+    private long createSegment(long id, TestContext context) {
         String name = getSegmentName(id);
         context.metadata.mapStreamSegmentId(name, id);
         initializeSegment(id, context);
@@ -1402,7 +1457,7 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
         return transactionId;
     }
 
-    private String getSegmentName(int id) {
+    private String getSegmentName(long id) {
         return "Segment_" + id;
     }
 
@@ -1410,6 +1465,18 @@ public class ContainerReadIndexTests extends ThreadPooledTestSuite {
         UpdateableSegmentMetadata metadata = context.metadata.getStreamSegmentMetadata(segmentId);
         metadata.setLength(0);
         metadata.setStorageLength(0);
+    }
+
+    private void markInactive(long segmentId, TestContext context) {
+        ((StreamSegmentMetadata) context.metadata.getStreamSegmentMetadata(segmentId)).markInactive();
+    }
+
+    private void evict(long segmentId, TestContext context) {
+        val candidates = Collections.singleton((SegmentMetadata) context.metadata.getStreamSegmentMetadata(segmentId));
+        val em = (EvictableMetadata) context.metadata;
+        val sn = context.metadata.getOperationSequenceNumber() + 1;
+        context.metadata.removeTruncationMarkers(sn);
+        em.cleanup(candidates, sn);
     }
 
     //endregion

